Can I use malloc?

I am using malloc with TinyGo but with nottinygc it panics.

How can I resolve this?

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x1199875]
```


`// These function are exported by TinyGo
	malloc := mod.ExportedFunction("malloc")
	free := mod.ExportedFunction("free")

	// Allocate Memory
	results, err := malloc.Call(ctx, nameSize)
	if err != nil {
		log.Panicln(err)
	}
	namePosition := results[0]

	// This pointer is managed by TinyGo,
	// but TinyGo is unaware of external usage.
	// So, we have to free it when finished
	defer free.Call(ctx, namePosition)
`

--- a/bench/bench_test.go
+++ b/bench/bench_test.go
@@ -13,6 +13,35 @@ import (
 	"github.com/tetratelabs/wazero/imports/wasi_snapshot_preview1"
 )
 
+// TestExportsMalloc is in the bench package since it already provides a fully packaged binary.
+// This could be cleaned up by using a different package, but for now we'll stick with being a little
+// lazy.
+func TestExportsMalloc(t *testing.T) {
+	wasm, err := os.ReadFile(filepath.Join("..", "build", "bench.wasm"))
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	ctx := context.Background()
+	r := wazero.NewRuntime(ctx)
+	wasi_snapshot_preview1.MustInstantiate(ctx, r)
+
+	mod, err := r.InstantiateModuleFromBinary(ctx, wasm)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	malloc := mod.ExportedFunction("malloc")
+	free := mod.ExportedFunction("free")
+
+	if malloc == nil {
+		t.Error("malloc is not exported")
+	}
+	if free == nil {
+		t.Error("free is not exported")
+	}
+}
+
 func BenchmarkGC(b *testing.B) {
 	tests := []string{"bench.wasm", "benchref.wasm"}
 	for _, tc := range tests {

--- a/init.go
+++ b/init.go
@@ -1,6 +1,8 @@
 // Copyright wasilibs authors
 // SPDX-License-Identifier: MIT
 
+//go:build tinygo
+
 package nottinygc
 
 /*

--- a/init.go
+++ b/init.go
@@ -4,6 +4,6 @@
 package nottinygc
 
 /*
-#cgo LDFLAGS: -Lwasm -lgc -lmimalloc -lclang_rt.builtins-wasm32
+#cgo LDFLAGS: -Lwasm -lgc -lmimalloc -lclang_rt.builtins-wasm32 --export=malloc --export=free
 */
 import "C"

--- a/magefiles/magefile.go
+++ b/magefiles/magefile.go
@@ -77,8 +77,6 @@ func Format() error {
 	return nil
 }
 
-var errMissingCopyrightHeaders = errors.New("missing copyright headers, use go run mage.go format")
-
 func Lint() error {
 	if _, err := sh.Exec(map[string]string{}, io.Discard, io.Discard, "go", "run", fmt.Sprintf("github.com/google/addlicense@%s", verAddLicense),
 		"-check",
@@ -89,7 +87,7 @@ func Lint() error {
 		"-ignore", "**/*.yml",
 		"-ignore", "**/*.yaml",
 		"."); err != nil {
-		return errMissingCopyrightHeaders
+		return fmt.Errorf("missing copyright headers, use go run mage.go format: %w", err)
 	}
 
 	return sh.RunV("go", "run", fmt.Sprintf("github.com/golangci/golangci-lint/cmd/golangci-lint@%s", verGolancCILint), "run")

--- a/magefiles/magefile.go
+++ b/magefiles/magefile.go
@@ -40,6 +40,14 @@ func Test() error {
 		return fmt.Errorf("unexpected error message: %s", s)
 	}
 
+	if err := buildBenchExecutable(); err != nil {
+		return err
+	}
+
+	if err := sh.RunV("go", "test", "./bench"); err != nil {
+		return err
+	}
+
 	return nil
 }
 
@@ -112,7 +120,7 @@ func UpdateLibs() error {
 
 // Bench runs benchmarks.
 func Bench() error {
-	if err := os.MkdirAll("build", 0o755); err != nil {
+	if err := buildBenchExecutable(); err != nil {
 		return err
 	}
 
@@ -120,11 +128,15 @@ func Bench() error {
 		return err
 	}
 
-	if err := sh.RunV("tinygo", "build", "-gc=custom", "-tags=custommalloc", "-scheduler=none", "-target=wasi", "-o", "build/bench.wasm", "./bench"); err != nil {
+	return sh.RunV("go", "test", "-bench=.", "-benchtime=10s", "./bench")
+}
+
+func buildBenchExecutable() error {
+	if err := os.MkdirAll("build", 0o755); err != nil {
 		return err
 	}
 
-	return sh.RunV("go", "test", "-bench=.", "-benchtime=10s", "./bench")
+	return sh.RunV("tinygo", "build", "-gc=custom", "-tags=custommalloc", "-scheduler=none", "-target=wasi", "-o", "build/bench.wasm", "./bench")
 }
 
 var Default = Test

